cassovary-core/src/main/scala/com/twitter/cassovary/graph/Traverser.scala

```diff
@@ -442,9 +448,67 @@ class BreadthFirstTraverser[+V <: Node](val graph: Graph[V], val dir: GraphDir,
     super.visitNode(node)
   }
 
-  def shouldBeEnqueued(color: Walk.NodeColor.Color): Boolean = {
-    color == Walk.NodeColor.Unenqueued
+  def shouldBeEnqueued(color: Color): Boolean = color == Walk.NodeColor.Unenqueued
+}
+
+/**
+ * Traverses in BFS order and keeps track of all shortest paths from any node in `homeNodeIds` to
+ * all target paths.  The traversal happens along all outbound nodes, it is suggested that `graph`
+ * storage direction is `GraphDir.OutDir`
+ * @param graph The graph to traverse on.  This graph need not necessarily be directed, but the
+ *              traversal will happen along all outbound nodes.
+ * @param homeNodeIds The sequence of source nodes to find all shortest paths.  Each shortest path must
+ *                    start with one of the home nodes.
+ * @param dir The direction of the graph traversal
+ * @param walkLimits  How the graph traversal will be limited
+ * @tparam V A subtype of `Node`
+ */
+class ShortestPathsTraverser[+V <: Node](val graph: Graph[V], val homeNodeIds: Seq[Int],
+                                        val dir: GraphDir = GraphDir.OutDir,
+                                        walkLimits: Walk.Limits = Walk.Limits())
+  extends DepthTracker with BoundedIterator[V] {
+
+  // Queueing parameters
+  override def nodePriority = GraphTraverserNodePriority.FIFO
+  override lazy val limits = walkLimits
+
+  // Accounting parameters
+  val numPaths = mutable.Map.empty[Int, Int]
+  val paths    = mutable.Map.empty[Int, Seq[Seq[Int]]]
+  homeNodeIds foreach { h => paths(h) = Seq(Seq()) }
+
+  /**
+   * To find all shortest paths, the chosen enqueued nodes must be selected carefully.
+   * In the filtering process, the shortest paths will be updated simultaneously.
+   * @return The next set of nodes to be enqueued
+   */
+  override protected def chooseNodes(node: Node): Seq[Int] = {
+    numPaths(node.id) = if (numPaths.contains(node.id)) numPaths(node.id) + 1 else 1
+
+    val currDepth = depth(node.id).get
+    val nextDepth = currDepth + 1
+
+    val index = numPaths(node.id) - 1
+
+    node.neighborIds(dir) filter { out =>
+      val shortestPathLength = depth(out)
+      val update = paths(node.id)(index) ++ mutable.Seq(node.id)
+
+      if (shortestPathLength.isDefined) {
+        if (nextDepth <= shortestPathLength.get) {
+          paths(out) = paths(out) :+ update
+          true
+        } else false
+      }
+      else {
+        paths(out) = mutable.Seq(update)
+        true
+      }
+    }
   }
+
+  //Always enqueue node independent of color
+  override def shouldBeEnqueued(color: Color): Boolean = true
 }
```

Comment:
what you have works, but is exponential in both time and space. What do you think about just keeping the information of 'seq of ancestors of a node'. So, in TestGraph.g6 if source node is 10. ancestors(10) = Nil, ancestors(11) = Seq(10), ancestors(12) = Seq(10), ancestors(13) = Seq(10), ancestors(14) = Seq(11, 12, 13), ancestors(15) = Seq(14). By the nature of BFS, I think it will work to not enqueue a neighbor n if it has been enqueued before, but append node.id to ancestors(n).

The full set of paths to a target node t will then be recursively the set of paths to ancestors(t) with 't' appended to each of them. This can be exponential but can be computed if/when needed by some other code, but at least not at the traverser level.